src/mesh/replicated_mesh.C

```diff
@@ -1417,5 +1417,151 @@ dof_id_type ReplicatedMesh::n_active_elem () const
+      {
+        list.push_back(elem);
+        interior_points.push_back(elem->centroid());
+        subdomains[elem->id()] = subdomain_counter;
+        break;
+      }
+    while (list.size() > 0)
+    {
+      // pop up an element
+      Elem * elem = list.back(); list.pop_back(); ++visited;
+
+      for (auto s : elem->side_index_range())
+      {
+        Elem * neighbor = elem->neighbor_ptr(s);
+        if (neighbor != nullptr && subdomains[neighbor->id()] == Elem::invalid_subdomain_id)
+        {
+          list.push_back(neighbor);
+          subdomains[neighbor->id()] = subdomain_counter;
+        }
+      }
+    }
+    std::cout << std::endl;
+    subdomain_counter++;
+  }
+  while (visited != n_elem());
+  // now subdomain_counter stores the number of disconnected subdomains
+
+  std::vector<std::vector<std::vector<Point>>> boundary_points(subdomain_counter);
+
+  // following boundary visiting algorithm only works with 2D
+  if (mesh_dimension() != 2)
+    libmesh_error_msg("Error: get_boundary_points only works for 2D now");
+
+  // get all boundary sides
+  // use a comparison functor to avoid run-time randomness due to pointers
+  struct boundary_side_compare
+  {
+    bool operator()(const std::pair<const Elem *, unsigned int> & lhs,
+                    const std::pair<const Elem *, unsigned int> & rhs) const
+    {
+      if (lhs.first->id() < rhs.first->id())
+        return true;
+      else if (lhs.first->id() == rhs.first->id())
+      {
+        if (lhs.second < rhs.second)
+          return true;
+      }
+      return false;
+    }
+  };
+  std::set<std::pair<const Elem *, unsigned int>, boundary_side_compare> boundary_elements;
+  for (const auto & elem : active_element_ptr_range())
+    for (auto s : elem->side_index_range())
+      if (elem->neighbor_ptr(s) == nullptr)
+        boundary_elements.insert(std::pair<const Elem *, unsigned int>(elem, s));
+
+  while (!boundary_elements.empty())
+  {
+    // get the first entry as the seed
+    const Elem * eseed = boundary_elements.begin()->first;
+    unsigned int sseed = boundary_elements.begin()->second;
+
+    // get the subdomain ID that these boundary sides attached to
+    subdomain_id_type subdomain_id = subdomains[eseed->id()];
+
+    // start visiting the mesh to find all boundary node with the seed
+    std::vector<Point> bpoints;
+    const Elem * elem = eseed;
+    unsigned int s = sseed;
+    std::vector<unsigned int> local_side_nodes = elem->nodes_on_side(s);
+    while (true)
+    {
+      boundary_elements.erase(std::pair<const Elem *, unsigned int>(elem, s));
+
+      // push all nodes on the side except the last one
+      for (unsigned int i = 0; i < local_side_nodes.size() - 1; ++i)
+        bpoints.push_back(*dynamic_cast<const Point *>(elem->node_ptr(local_side_nodes[i])));
+
+      // use the last node to find next element and side
+      const Node * node = elem->node_ptr(local_side_nodes[local_side_nodes.size() - 1]);
```

Comment:
I don't think this is going to work on second-order elements.  On Quad8/Quad9/Tri6, the last index nodes_on_side returns is actually the local index of the mid-edge node, not of either of the corners.  Maybe go in the opposite direction?  In theory we haven't promised this, but in practice local_side_nodes[0] will always be a vertex.  Or if direction is important then at least *currently* local_side_nodes[1] is always the other vertex in 2D or another vertex in 3D.

Reply:
And now that I'm done reading back and forth, I think this is the only place where either I'm really confused or there actually is a bug.  All the other changes I've suggested are either minor tweaks, or at worst are debatable API design decisions.

Reply:
This is the place I am not certain as well. It will be nice if we can do some unit test on this function, but I do not know how to do it. I think it is probably worthwhile to come up with something.